appinventor/appengine/src/com/google/appinventor/server/storage/ObjectifyStorageIo.java

```diff
@@ -2025,36 +2026,59 @@ public ProjectSourceZip exportProjectSourceZip(final String userId, final long p
         @Override
         public void run(Objectify datastore) throws IOException {
           Key<ProjectData> projectKey = projectKey(projectId);
-          boolean foundFiles = false;
           for (FileData fd : datastore.query(FileData.class).ancestor(projectKey)) {
+            fileData.add(fd);
+          }
+          for (int i = 0; i < fileData.size(); i ++) {
+            FileData fd = fileData.get(i);
+            String fileName = fd.fileName;
+            if (fileName.startsWith("src/") && (fileName.endsWith(".scm") || fileName.endsWith(".bky") || fileName.endsWith(".yail"))) {
+              String fileNameNoExt = fileName.substring(0, fileName.lastIndexOf("."));
+              int count = screens.containsKey(fileNameNoExt) ? screens.get(fileNameNoExt) + 1 : 1;
+              screens.put(fileNameNoExt, count);
+            }
+          }
+          Iterator it = fileData.iterator();
+          while (it.hasNext()) {
+            FileData fd = (FileData)it.next();
             String fileName = fd.fileName;
             if (fileName.startsWith("assets/external_comps") && forGallery) {
               throw new IOException("FATAL Error, external component in gallery app");
             }
-            if (fd.role.equals(FileData.RoleEnum.SOURCE)) {
+            if (!fd.role.equals(FileData.RoleEnum.SOURCE)) {
+              it.remove();
+            }
+            else {
               if (fileName.equals(FileExporter.REMIX_INFORMATION_FILE_PATH)) {
                 // Skip legacy remix history files that were previous stored with the project
-                continue;
+                it.remove();
               }
               if (fileName.startsWith("screenshots") && !includeScreenShots) {
                 // Only include screenshots if asked...
-                continue;
+                it.remove();
+              }
+              if (fileName.startsWith("src/") && (fileName.endsWith(".scm") || fileName.endsWith(".bky") || (fileName.endsWith(".yail") && includeYail))) {
+                String fileNameNoExt = fileName.substring(0, fileName.lastIndexOf("."));
+                if ((Integer)screens.get(fileNameNoExt) < 3) {
+                  LOG.log(Level.INFO, "Not adding file to build ", fileName);
+                  it.remove();
+                  deleteFile(userId, projectId, fileName);
```

Comment:
I wonder if we could somehow kick this off as a separate job just so it doesn't interfere with the remainder of the export process if it fails for some reason. @jisqyv Any thoughts?